sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIOTest.java

```diff
@@ -0,0 +1,315 @@
+    PCollection<String> output =
+        raw.apply(
+            MapElements.into(TypeDescriptors.strings())
+                .via(
+                    (RabbitMqMessage message) ->
+                        new String(message.getBody(), StandardCharsets.UTF_8)));
+
+    List<String> records =
+        generateRecords(maxNumRecords)
+            .stream()
+            .map(record -> new String(record, StandardCharsets.UTF_8))
+            .collect(Collectors.toList());
+    PAssert.that(output).containsInAnyOrder(records);
+
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = null;
+    Channel channel = null;
+    try {
+      connection = connectionFactory.newConnection();
+      channel = connection.createChannel();
+      channel.exchangeDeclare("READEXCHANGE", "fanout");
+      Channel finalChannel = channel;
+      Thread publisher =
+          new Thread(
+              () -> {
+                try {
+                  Thread.sleep(5000);
+                } catch (Exception e) {
+                  LOG.error(e.getMessage(), e);
+                }
+                for (int i = 0; i < 10; i++) {
+                  try {
+                    finalChannel.basicPublish(
+                        "READEXCHANGE",
+                        "test",
+                        null,
+                        ("Test " + i).getBytes(StandardCharsets.UTF_8));
+                  } catch (Exception e) {
+                    LOG.error(e.getMessage(), e);
+                  }
+                }
+              });
+      publisher.start();
+      p.run();
+      publisher.join();
+    } finally {
+      if (channel != null) {
+        channel.close();
+      }
+      if (connection != null) {
+        connection.close();
+      }
+    }
+  }
+
+  @Test
+  public void testWriteQueue() throws Exception {
+    final int maxNumRecords = 1000;
+    List<RabbitMqMessage> data =
+        generateRecords(maxNumRecords)
+            .stream()
+            .map(bytes -> new RabbitMqMessage(bytes))
+            .collect(Collectors.toList());
+    p.apply(Create.of(data))
+        .apply(
+            RabbitMqIO.write().withUri("amqp://guest:guest@localhost:" + port).withQueue("TEST"));
+
+    final List<String> received = new ArrayList<>();
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = null;
+    Channel channel = null;
+    try {
+      connection = connectionFactory.newConnection();
+      channel = connection.createChannel();
+      channel.queueDeclare("TEST", true, false, false, null);
+      Consumer consumer = new TestConsumer(channel, received);
+      channel.basicConsume("TEST", true, consumer);
+
+      p.run();
+
+      while (received.size() < maxNumRecords) {
+        Thread.sleep(500);
```

Comment:
There is no problem with this sleep because it is not a wait for an absolute period of time but more a way to temporise the check of the receive queue depth. As a very minor suggestion (only if you feel the envy :) ), I would implement it with a new CountDownLatch(maxNumRecords); countdown.await() and countdown.countDown in the TestConsumer when it receives a message.
But once again it is very minor way to reduce test time a tiny bit but I'm ok with current impl